webpack/assets/javascripts/react_app/routes/common/reducerHOC/withDataReducer.js

```diff
@@ -0,0 +1,36 @@
+import Immutable from 'seamless-immutable';
+
+const initialState = Immutable({
+  isLoading: true,
+  isFetching: false,
```

Comment:
I'm not sure I understand the need for separate `isLoading` and `isFetching` states. 
Could you please explain what does it mean when:
1. `isLoading == true && isFetching == false` 
2. `isLoading == false && isFetching == true`
(I guess the other two combinations are fairly obvious)

Reply:
I'm still in the thought process. I want to separate the initial loading of the component and the loading of fetching new data. but when I think about it now `isLoading` is enough, we just need to update `withRenderHandler` to show a spinner/skeleton only when `hasData` is false.

if `hasData` is false ( empty or error should change it to false ) and `isLoading` is true, we should display a skeleton/spinner. e.g. in a table view a search query returned empty results, so there is no table anymore, only a message that says empty. when the user searches again he should see a skeleton before the results arrive.

if `hasData` and `isLoading` are true, we shouldn't show a skeleton/spinner, we should just show the current data until the new data arrives.

Reply:
@sharvit can you ack?

Reply:
@glekner I think we should try and understand the possible states here and what are the possible transitions between them, before deciding which flags to use to represent them. 
My initial thought, and I may be wrong about this, is that a component using this can be in one of three states:
1. Didn't load yet because it is waiting for a response - this is the initial state, and can transition to either state 2. or 3. depending on the result of the server response
2. Loaded correctly - this can transition back to state 1 if the user triggers some change that requires loading new data (e.g. changing table sort order)
3. Error during loading - either because the loading itself failed or because the server replied with an error response. This again can only transition back to state 1 if the user triggers something that loads new data.

If we only have 3 states, we shouldn't use 3 flags to represent them since there would be 5 combinations that "don't make sense" and might lead to bugs. if that is indeed the case, i'd suggest using just two flags - `hasData` and `hasError` - if both are false, it's loading. That still leaves one "useless" state of both true, but that's better imho. Or we could use an enum to represent only the actual states.

Reply:
I'll try to sum it up. I believe this covers all the cases? @tbrisker 

Views: **Loading**, **Message**(Empty or Error) and **Component**

| isLoading  | hasData | hasError | VIEW | Comments |
| ------------- | ------------- | ----------- | ------------ | ------------ |
| TRUE.  💚   | FALSE :red_circle:  | FALSE :red_circle: | **Loading** | Initial Loading **OR** after Message |
| FALSE :red_circle:  | TRUE 💚  | FALSE :red_circle: | **Component** | Show Data |
| TRUE. 💚  | TRUE 💚  | FALSE :red_circle: | **Component** | Outdated data exists, should not show loading to avoid flickering so we still show the component with the outdated data until the new data arrives |
| FALSE :red_circle:  | FALSE :red_circle:  | FALSE :red_circle: | **Message** | query returned 0 results, this is not an error, and we display a message |
| FALSE :red_circle:  | FALSE :red_circle:  | TRUE 💚 | **Message** | query failed, and we display a message |

should also mention that triggering an API call resets `hasError` to false.

Reply:
Thanks, much clearer now. I'm not sure if 2 and 4 shouldn't be the same state and just have the component decide what to do depending on eg. data.length, but i can live with this as well.

Reply:
We want to avoid conditional rendering inside the component, `withRenderHandler` takes care of it now :)